### PR TITLE
Support version 1.0 for openfe and gufe 

### PR DESCRIPTION
**Pull request by ijpulidos**
This PR supersedes #33. 

Solves #31 
Solves #22 

It aims to add the following changes in order to support the 1.0 version/rc of `gufe` and `openfe`.

TODO:
- [x] Fix IntegratorSettings
  - [x] Make a new integrator settings class for NEQ integrator
- [x] Fix solvation settings
  - [x] _should_ be just an import replacement
- [ ] Deal with mapping changes
  - [ ] SIngle mapping validation
  - [ ] Make sure everything is now a list rather than a dictionary
  - [x] Remove dependency on "ligand" dictionary entry in mapping
- [x] LJ Alchemical settings

### CODE REVIEW

File: feflow/settings/integrators.py
Comment (IAlibay) [inline feflow/settings/integrators.py:25]: Sorry very much a flyby 2 second review for now - the IntegratorSettings need to include information about the barostat frequency & possibly com removal, see: https://github.com/OpenFreeEnergy/openfe/blob/c4678745900e6425354407de5885c7c3eefd429b/openfe/protocols/openmm_utils/omm_settings.py#L195-L201

I'm not particularly sure it's the right design pattern, see: https://github.com/OpenFreeEnergy/openfe/issues/712, however because `get_system_generator` (https://github.com/OpenFreeEnergy/openfe/blob/main/openfe/protocols/openmm_utils/system_creation.py#L26) currently asks for an integrator settings object (and you're using it in the neq protocol), you end up needing that additional field.

Comment (IAlibay) [reply feflow/settings/integrators.py:25]: Ah maybe there's going to be a typing issue - I can see that maybe we have to have a discussion about this.

File: feflow/settings/integrators.py
Comment (IAlibay) [inline feflow/settings/integrators.py:31]: Usually we put this under "SimulationSettings"

Comment (ijpulidos) [reply feflow/settings/integrators.py:31]: In this case these are specific to the integrator as a single cycle/run of the integrator does a "equilibration on endstate A" -> "nonequilibrium sim on A to B (changing lambda)" -> "equilibration on endstate B"  -> "nonequilibrium sim on B to A". So these parameters are to change how many steps on each of the eq/neq parts. I hope that makes sense.

Comment (IAlibay) [reply feflow/settings/integrators.py:31]: I see what you mean. This is a bit of an odd once since, as a user, I would expect these to be more "how I define my simulation protocol" than "something related to my integrator". It probably doesn't matter, but it's something to consider longer term.

File: feflow/settings/integrators.py
Comment (ijpulidos) [reply feflow/settings/integrators.py:25]: The `barostat_frequency` doesn't seem to be needed or used, as far as I can tell, at least not in the NeqCycling protocol, which makes me wonder if this is something we would like to change. What are the scenarios where changing this frequency is desired?

**Comment by ijpulidos** [comment]
This one should be now pretty much now ready to go.

The failing tests are the ones with "bad" mappings. Now the `LigandAtomMapping` implementation checks for boundaries, so the way we were forcing a "failed dag execution" is no longer working. We would have to think about better ways to test a failed dag execution (I think it's desired to have a "negative" test). Is the `openfe` package doing a similar test?

On the other hand, we should probably wait for a 1.0 release of both `gufe` and `openfe` before merging this into `main`.

**Comment by IAlibay** [comment]
Awesome your work here @ijpulidos ! I'll try to have a look tomorrow morning ahead of our call.

File: feflow/settings/integrators.py
Comment (ijpulidos) [reply feflow/settings/integrators.py:25]: In the end it was indeed needed for "real-life" simulations, this probably means we need to add more testing that resembles this situation.

File: feflow/settings/integrators.py
Comment (hannahbaumann) [inline feflow/settings/integrators.py:30]: Is the default here 250000 (steps?) or 12500?

File: feflow/settings/nonequilibrium_cycling.py
Comment (IAlibay) [inline feflow/settings/nonequilibrium_cycling.py:71]: In the next iteration of this work we should switch this to use the OutputSettings and EngineSettings we have elsewhere, that way the user API is more consistent.

@dotsdl how big of a deal would it be if we did this in the next release? (i.e. would we need to do some kind of migration?)

**Review by IAlibay — CHANGES_REQUESTED** [review]
This looks good for a first pass.

Things that aren't included here but would improve science (medium-high priority):
  - Partial charge settings & fixing partial charge setting

Things that would make things easier for future-ability (medium priority unless @dotsdl feels otherwise):
  - Make the Protocol an input of SetupUnit not the Settings (this would fix migration issues)
  - Settle down on the Settings to avoid changing them further / migrations

Things that would improve usability / reduce errors (low priority):
  - Add more guardrails on mappings & general end state validation

**Comment by dotsdl** [comment]
@IAlibay I will add the partial charge setting (and *not* setting) approach used in the OpenFE `RelativeHybridTopologyProtocol`; that will be critical for users of `alchemiscale-fah`.

**Comment by dotsdl** [comment]
@IAlibay can you work with @ijpulidos to get the settings changes that you have identified into this PR? That will avoid painful migrations later on, especially if we already know the changes we want made.

File: feflow/protocols/nonequilibrium_cycling.py
Comment (ijpulidos) [reply feflow/protocols/nonequilibrium_cycling.py:129]: Hmm, I see what you mean. I think that's a good option, but I wonder if that hinders the possibility of using this Unit in other protocols that do not necessarily rely on ligands mappings (for example in a protein-mutation protocol). That is, assuming we want to try to reuse the units and potentially not having redundant code to do similar things in each protocol.

File: feflow/settings/nonequilibrium_cycling.py
Comment (ijpulidos) [reply feflow/settings/nonequilibrium_cycling.py:71]: One thing with the `OutputSettings` is that the current state is pretty oriented towards the repex protocol, and the use of the `openmmtools.multistate.MultiStateSampler`, so it's not that transferable to the Neq Cycling protocol. We probably need to write some NEq sampler on `openmmtools`  that hopefully shares some API with the `MultiStateSampler` but this seems like a non-trivial amount of work.

Using the `OpenMMEngineSettings` should be already possible. I'll make the changes for that.

**Comment by ijpulidos** [comment]
As discussed in the dev meetings, I think we should aim this PR to be the minimal set of changes that would support the released 1.0 versions of openfe and gufe. This is just a convenient way to be able to handle the scope of each PR in a more granular way and avoid trying to change too much in each.

I have created issues for the remaining changes suggested in https://github.com/choderalab/feflow/pull/38#pullrequestreview-2093798032 and added them to the [next milestone](https://github.com/choderalab/feflow/milestone/1)

This should be ready for a new review, considering some of the suggestions in previous reviews are to be solved in subsequent PRs, as discussed.

**Comment by ijpulidos** [comment]
Please note that the CI tests are failing due to the temporary drop of support for GAFF (that's also why it's succeeding for older python 3.9 version). These fails should be independent of the set of changes in this PR as far as I can tell.

File: feflow/settings/nonequilibrium_cycling.py
Comment (hannahbaumann) [inline feflow/settings/nonequilibrium_cycling.py:80]: Maybe also change the doc string to reflect the name change

**Review by hannahbaumann — COMMENTED** [review]
Just the small doc string suggestion, otherwise looks great to me!

File: feflow/settings/nonequilibrium_cycling.py
Comment (IAlibay) [reply feflow/settings/nonequilibrium_cycling.py:74]: I'll have a think - I think we can do make it so that the "look and feel" is the same, even if the two protocols differ in terms of base classes.

File: feflow/protocols/nonequilibrium_cycling.py
Comment (IAlibay) [reply feflow/protocols/nonequilibrium_cycling.py:129]: that's a good point - long term maybe the type check should be that the "Protocol" level rather than the "Unit" level?

short term, I generally take the approach of "tight checks on untested features" and then loosen up later

Anyways, this is not something to deal with here, maybe in a next iteration.

**Review by IAlibay — APPROVED** [review]
this lgtm! it's a good starting point for following up on a few things later

File: feflow/protocols/nonequilibrium_cycling.py
Comment (ijpulidos) [reply feflow/protocols/nonequilibrium_cycling.py:129]: I've thought about this a little bit more and yes, you are right. While I'd love to avoid redundant code it is something that will happen when we tackle implementing other protocols and integrating them to this code base. I agree. I'll change the type to be a `LigandAtomMapping` as you suggested, it shouldn't affect things.

**Pull request merged**
